### ISSUE TITLE
Run Rails in the GitHub Actions Container for CI

### ISSUE THREAD

**Issue by managedkaos** [EBWiki/EBWiki#3903]
Presently, some CI workflows build and run a monolithic container to provide a live environment (rails server and database).

This causes the workflow to have an extended run time and, on occasion, the docker process does not start as expected.

It's possible that the live environment could be run inside the GitHub Actions container where the workflow is running.  That is, start Rails and Postgres services in the CI container instead of building and running a container _inside_ the CI container.

The following documents provide examples for this approach.  Review the documents and develop an implementation that will support the needs of the EBWiki project.

- [Building a Rails CI pipeline with GitHub Actions
](https://boringrails.com/articles/building-a-rails-ci-pipeline-with-github-actions/)
- [GitHub Actions with Ruby on Rails: Setting up Continuous Integration
](https://gorails.com/episodes/github-actions-continuous-integration-ruby-on-rails)
- [Github Actions for Ruby on Rails CI
](https://medium.com/@OwenTran/github-workflow-for-rails-ci-34209a53d19e)
- [Simple use of github action with Rails project](https://github.com/niclin/rails-github-actions-example)
- [Github Actions to run Rubocop and RSpec tests on Rails with Postgres
](https://dev.to/abdellani/github-actions-to-run-rubocop-and-rspec-tests-on-rails-with-postgres-47i)

**Comment by mockdeep** [comment]
@managedkaos I think it'll probably make a pretty big difference if we run the app on the CI container directly. I'm happy to tackle this project. I think this might be a good time to bring up my comments [from another issue](https://github.com/EBWiki/EBWiki/issues/3709#issuecomment-770334878) about switching to CircleCI:

I've been using Github Actions on some projects and CircleCI on others, and I'm still finding CircleCI to be the better option. Benefits:

* Builds seem to be consistently faster and more reliable. I've had a number of situations where the build hangs after 10 minutes on Github Actions. I've even seen a couple build hangs on my PRs on EBWiki.
* CircleCI has a nicer UI. It's pretty awkward to get to the actual failure on Github Actions. CircleCI focuses the failure and gives you a nice description if you use their test reporter.
* All steps run on CircleCI. On Github Actions it kills ongoing build steps when one step fails. This is often frustrating when you're trying to get a better picture of exactly which tests are failing and why.
* Parallelization. CircleCI has some great parallelization tools. The build runs pretty fast right now, but on CircleCI you can easily split builds up into as many as 80 separate jobs.

If you're game, I'm pretty familiar with CircleCI and can set it up for EBWiki.

**Comment by managedkaos** [comment]
Appreciate the input.

As for switching from Actions to CircleCI, I think we're set on Actions for the time being.  Historically, the project was running on TravisCI.  My last major project for EBWiki was moving from Travis to Actions and we're just getting into the groove of running here.  I'd prefer not to disrupt that for now.

Also, I appreciate having the majority of the workflow in one place.  Our contributors are here, the code is here, and it just helps that the CI is here as well.  I don't think Actions is the end-all-be-all of CICD but for this project, it just kinda works.

Also, consider the cost and management overhead of using a paid CI system for a non-profit and volunteer-run organization.  The free tier of CircleCI provides 2500 credits/week with the ability to run 1 job at a time.  If we have a few jobs, they'd be parallelized which may slow things down a bit.  I don't think we'd hit the credits limit regularly; perhaps during heavy contribution times like Hacktoberfest.  The compute is free on Actions, albeit a bit slow and perhaps unstable at times.  But given the price that we are paying and the low management overhead, I'm OK with that.

**Comment by mockdeep** [comment]
Open source projects actually [get 400k free minutes](https://circleci.com/open-source/) per month on CircleCI, so it seems really unlikely we'd get anywhere close to that. They also allow running [4 concurrent builds](https://circleci.com/docs/2.0/oss/index.html). Build pages for open source are public (e.g.: [React](https://app.circleci.com/pipelines/github/facebook/react)), so I don't think there's any meaningful difference as to whether they are in one place or two.

Like I said, I'm happy to do the legwork getting it set up, and it's pretty easy to get it running on the side without blocking anybody's workflow. Having to page down through dozens of pages of output on Github Actions in order to find the actual error is kind of maddening.

One additional benefit of CircleCI is being able to SSH into builds. I've found that really handy when I'm unable to reproduce an issue locally.

**Comment by mockdeep** [comment]
Actually, here's what I can do. I'll set up both so we have a better comparison point.

**Comment by mockdeep** [comment]
Okay, I've got both of them set up and running. [Here's Github Actions](https://github.com/EBWiki/EBWiki/pull/3927/checks) and [here's CircleCI](https://github.com/EBWiki/EBWiki/pull/3906). Both builds are 2-3 minutes.

**Comment by managedkaos** [comment]
Looks good to me!

With Actions running the way it is, I'm more inclined to move forward with #3927 since we already have Actions set up.

Can you share a screencap image/pdf of the CircleCI run?  I'd rather not auth it to my account at the moment.

**Comment by mockdeep** [comment]
@managedkaos I've added some screen grabs and descriptions [over here](https://github.com/EBWiki/EBWiki/pull/3906#issuecomment-776140306).

> since we already have Actions set up.

I've already done the legwork on CircleCI, so it would require < 5 minutes of work for a repo admin to switch it over. We'd need to 1. enable it on the main repo (right now it's on my fork) and 2. configure the branch protections to require CircleCI instead.

**Comment by managedkaos** [comment]
I'll leave it up to the project maintainers to decide if they want to switch to Circle CI.

But again, I'm inclined to believe they will want to stick with Github Actions for the time being.  With this being an open source, volunteer effort, it's a concern to minimize the sprawl of the resources used to develop the application.  There's already Heroku, AWS, Github, and others.  If the CI for the app can be centralized in one of those accounts/tools already in place, that's one less account that the maintainers have to worry about.

I say this not to bang on Circle CI; I think it's a great tool and your experiment shows that it can definitely be applied to EBWiki's development.  But you've also developed a solution that cuts the current integration time in Github Actions down by 4x!! That's awesome.  I'm saying we run with those improvements for now and see how much mileage we can get out of that.

**Comment by mockdeep** [comment]
@trystant @rlgreen91 do you two want to weigh in on this one?

**Comment by rlgreen91** [comment]
Yeah, let's stick with Github Actions.   We literally switched over from TravisCI to Github Actions last fall - there's no reason to leave it now if it works for our needs, and it does.  There's really just not a compelling reason to move to CircleCI - maybe at some point we'll need the parallelization options but that's not the case right now.

It sounds like I can just go ahead and merge #3927, yes?  If not, ping me here.

**Comment by mockdeep** [comment]
Okay, I've marked the Github Actions PR as ready for review. @rlgreen91 if there are branch protections set up, you'll need to adjust them to match the build on #3927.

**Comment by mnyon-grandkru** [comment]
@mockdeep @managedkaos @rlgreen91 One thing I'm curious about; what are the runtime comparisons between CircleCI & Github actions? Having an idea of what the differential will make a potential future switchover (or disaster recovery/service failure switchover) relatively easy.

**Comment by mockdeep** [comment]
@mnyon-grandkru the run times were pretty similar. Github Actions was ~2 minutes and CircleCI was ~3 minutes. The extra minute was aggregating test coverage results, which the Github Actions setup does not do.

**Comment by managedkaos** [comment]
I'm pretty sure we can close this issue given the [CI](https://github.com/EBWiki/EBWiki/actions/workflows/ci.yml) workflow is now in place.

I'll plan to close this issue in the new few days unless I hear anything different. 👍
